wallet/txauthor/author.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2016 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+// Package txauthor provides transaction creation code for wallets.
+package txauthor
+
+import (
+	"errors"
+
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/btcsuite/btcwallet/wallet/txrules"
+
+	h "github.com/btcsuite/btcwallet/internal/helpers"
+	"github.com/btcsuite/btcwallet/wallet/internal/txsizes"
+)
+
+// InputSource provides transaction inputs referencing spendable outputs to
+// construct a transaction outputting some target amount.  If the target amount
+// can not be satisified, this can be signaled by returning a total amount less
+// than the target or by returning a more detailed error implementing
+// InputSourceError.
+type InputSource func(target btcutil.Amount) (total btcutil.Amount, inputs []*wire.TxIn, scripts [][]byte, err error)
+
+// InputSourceError describes the failure to provide enough input value from
+// unspent transaction outputs to meet a target amount.  A typed error is used
+// so input sources can provide their own implementations describing the reason
+// for the error, for example, due to spendable policies or locked coins rather
+// than the wallet not having enough available input value.
+type InputSourceError interface {
+	error
+	InputSourceError()
```

Comment:
It was described in the comment.  It's so the caller can choose to return a more detailed error if they want while allowing easy detection of an `InputSourceError` versus some other type of error such as a database failure, etc.  Check out how the tests switch on the error type.

For example:

```Go
type CustomInputSourceError struct {
	SomeImportantField1 int
	SomeImportantField2 int
}

// Make sure it is identifiable by the caller as an `InputSourceError`.
func (c *CustomInputSourceError) InputSourceError() {}

// .... Later in the caller somewhere...

if _, ok := err.(txauthor.InputSourceError); ok {
	// Handle any type of input source error differently
	// regardless of whether it was a custom one or the
	// standard implementation provided by the package.
} else {
	// Some other type of unexpected error.  Handle it differently.
}
```